### DISCUSSION TITLE
Manual sorting of products

### DISCUSSION THREAD

**Discussion by andris-sevcenko** [craftcms/commerce#2302] [Ideas] [answered]
> Created by: Dominik ([dominik@flexify.net](dominik@flexify.net)) on 2016/02/23 10:40:57 +0000
Votes at time of UserVoice import: 61

------
Make products manually sortable

(e.g. to map the sort order to some kind of popularity metric)

**Comment by andris-sevcenko** [comment]
> Created by: Jeremy Daalder ([jd@imagescience.com.au](jd@imagescience.com.au)) on 2016/07/23 04:14:19 +0000


------
Products should be sortable by drag and drop in the back end, and when you pull products into templates this manual order should be the default and/or and an option (.order(&#39;manual&#39;) or similar)

**Comment by andris-sevcenko** [comment]
> Created by: Terry Upton ([terry@amasci.co.uk](terry@amasci.co.uk)) on 2016/07/23 12:40:15 +0000


------
This would be really useful. +1

**Comment by andris-sevcenko** [comment]
> Created by: Alex ([alex@sustainit.com](alex@sustainit.com)) on 2017/03/01 16:54:56 +0000


------
we need this!

**Comment by andris-sevcenko** [comment]
> Created by: Thijs Kaspers ([thijs@pixelcode.nl](thijs@pixelcode.nl)) on 2017/06/22 08:57:01 +0000


------
Yes! Like a structure, drag/drop order..

**Comment by andris-sevcenko** [comment]
> Created by: Peter Oliver Geller ([petergeller@googlemail.com](petergeller@googlemail.com)) on 2017/08/08 16:58:04 +0000


------
Is that feature planned in the next time, otherwise we try to do it differently.

**Comment by andris-sevcenko** [comment]
> Created by: Anonymous ([jacques.letesson@gmail.com](jacques.letesson@gmail.com)) on 2017/11/14 18:07:51 +0000


------
A must

**Comment by andris-sevcenko** [comment]
> Created by: nermin ([nermindk@yahoo.dk](nermindk@yahoo.dk)) on 2018/03/04 12:28:39 +0000


------
Definitely a must to have

**Comment by andris-sevcenko** [comment]
> Created by: Mat Johnson ([moldedjelly@gmail.com](moldedjelly@gmail.com)) on 2018/03/16 01:34:19 +0000


------
was thinking it could operate like entries - ie, structured vs channel

**Comment by terraindesign** [comment]
+1 This is a must have for client.

**Comment by matt-adigital** [comment]
+1 to this! Would be extremely useful if we could reorder products within a product type like we can with structure entries currently.

**Comment by lukeyouell** [comment]
+1 it's a common request from clients

**Comment by rob-c-baker** [comment]
+1 This would be extremely useful.

It's not a good user experience to have a custom, number based "sorting" field to achieve this.

**Comment by ghost** [comment]
+1 Must have and extremely useful!

**Comment by dougalcrowder** [comment]
+1 
Currently hacking a workaround, so would be great to have outta the box

**Comment by stenvdb** [comment]
Asked in Discord why this was removed from the 3.0 milestone and got a reply back from @nfourtythree (thanks by the way!). Thought this might interest others:

> @stenvdb that was removed from the 3.0 milestone due to some dependencies from Craft core. At the moment we would suggest using a custom number field to set an order, this along with an on save element event to check for new products and automatically setting the number in the custom field.
> 
> I understand that this isn't quite the same functionality in the CP as a structure but it will give you the ability to have manually sorted products.

Must say, being this is the most commented on issue here, it's a bummer. Will probably build some small module/plugin to account for the lack of this functionality. Can't afford to wait on a 3.x release to add this funct.

**Comment by nfourtythree** [comment]
@stenvdb there is also another way to do this by using a custom Product field.

If you create a global set and add a custom product field. Then select all the products. This gives you the ability to drag and drop the order of the products.

You can use this in conjunction with the on save event and catch the save of new products and programmatically add them to the field mentioned above.

**Comment by green17** [comment]
This would be really useful. +1

**Comment by denisyilmaz** [comment]
any updates on this? we are currently setting up a shop for a client and they need to have custom sorting of their products. would be great if we could use native sorting and don't need to rely on some custom fields for this.

**Comment by improved-software** [comment]
Would also appreciate some feedback on if this is currently in the works. 

A requirement for a client.

**Comment by ArtDepartmentMJ** [comment]
this would be a great feature

**Comment by simonfranzen** [comment]
Vote for this +1

**Comment by tonyastley** [comment]
This feature is pretty much a deal breaker with or client base. Would be great.

**Comment by billmn** [comment]
+1 A must to have

**Comment by chillibe** [comment]
+1

**Comment by Geestig** [comment]
Is there any update regarding which release you're targeting for this functionality?

**Comment by Elliot316** [comment]
Would absolutely benefit from a better sort function. With over 5,000 individual products, it's crucial to be able to organize product on PLPs similar to how you organize product on the PDP image slider, ie. a drag and drop. Right now we organize product by date to structure the way we want, which is so very tedious.

**Comment by solutiondrop** [comment]
This would be so useful!

**Comment by gazlawn22** [comment]
Yeah, big one for us too!

**Comment by RitterKnightCreative** [comment]
Yeh this just [came up](https://craftcms.stackexchange.com/questions/36622/craft-commerce-ordering-products-like-a-structure/36626#36626) again on the Craft Stack Exchange.

**Comment by terryupton** [comment]
Coming from Commerce 1 to Commerce 3 for a client and this is one of their 'Key' requirements/requests for the new website.
Is this something planned at all?

**Comment by nitech** [comment]
This is an important missing feature. It is kind of related to faceted navigation/filtering of products. Eagerly awaiting more wholesome set of presentation techniques for product lists.

**Comment by aeu** [comment]
This is definitely something we would want as well

**Comment by tigerchick** [comment]
We really need this too. Is there anything in the works?

**Comment by stolat** [comment]
Definitely need this for a current project too!

**Comment by adrienpicard** [comment]
+1 this is an important feature for clients

**Comment by sm9** [comment]
+1 for this feature. Clients have requested this feature a few times to us, and we've been told that they could previously manually re-order products in WooCommerce and Shopify, so this was something they had expected to be able to do in Craft Commerce. I don't see any sign of this feature in the Commerce 4 beta either, unfortunately. The workarounds for using custom ordering fields aren't ideal.

**Comment by Nellyaa** [comment]
+1 also need this :)

**Comment by splendidrob** [comment]
+1 this should be the same as other entries as per a regular structure - clients want to be able to manually order their products

**Comment by tigerchick** [comment]
Have you considered assigning categories to products and then controlling order through each category instead? 
I found this to be an effective solution - and arguably better than ordering products directly as it means that order can be controlled on a per category basis (which, granted, may or may not be a good idea depending on your frontend architecture).

**Comment by wmdhosting** [comment]
Variants now  can be reordered manually.. as Structure 

Prodcuts  types >>  by default work as >> **Channel type**   section

Can you add option to select they work as **Structure**..  too

**Comment by sandrodunkel** [comment]
I was hoping for this functionality in Commerce 5 but it's still not possible. There is a [workaround described in the Knowledge Base](https://craftcms.com/knowledge-base/manually-sorting-commerce-products), but I feel like this can't be the solution to this problem. Are there plans to implement this feature?

**Comment by masiorama** [comment]
Curious about this too. My clients ask for this feature, natively.

**Comment by sandrodunkel** [comment]
Issue #3630

**Comment by wmdhosting** [comment]
This is done in 5.2

**Reply by masiorama** [reply]
How exactly?

**Comment by wmdhosting** [comment]
Structure added .. in 5.2 for prodcuts same as in sections..

**Reply by lukeholder** [reply, marked as answer]
Since Craft Commerce 5.2 Product types can now be set to be a structure just like entries, so products are order-able and nest-able. 

Enjoy!

**Reply by masiorama** [reply]
@lukeholder @wmdhosting Checking inside product type the switch says: "Changing this could result in a loss of data.". What kinda of data? I mean, is it not safe to upgrade an already existing commerce catalog with hundreds of products?

**Reply by wmdhosting** [reply]
Thats normal precaution message...  
just to say user..  make DB backup just in case.. :)
But all works ok..

**Reply by masiorama** [reply]
@wmdhosting thanks. I'm testing on dev env: how am I supposed to control products order? Can you point me to the exact place where this feature is documented, please? Cannot find it in the docs.

**Reply by wmdhosting** [reply]
Its simple. after you switch.. 
you will get drag and drop  icon before status for every product
Dont know if its documented..as that failry simple. 
If no documneted.. look in Craftcms section structure option works in same way..

**Reply by masiorama** [reply]
I checked that but, in my case, I'm querying products (variants actually) and presenting in grid via _postDate DESC_ (product), _sortOrder ASC_ (variant). I cannot find the correct field to check, something like _productOrder_ to fix my _orderBy_ parameter.

...
Ok, maybe I got it right while searching the doc: _lft ASC_
I'm gonna test and report in a few minutes.

Updated: can't make it work.
My scenario is to grab all variants: `craft.variants.hasProduct(relatedProducts)` and then 

```
{% set results = allVariants
        // limit, search, relatedTo and with not relevant stuff
	.orderBy('featuredProduct DESC, postDate DESC, sortOrder ASC, score')
%}
```
... which works.
If i add `lft ASC` to order by the query is broken `Column not found: 1054 Unknown column 'lft' in 'order clause'`.

**Reply by wmdhosting** [reply]
In my case i use like this if helps anything 
   
```
    {% elseif productType|length == 1 %}

        {% set productTypeData = craft.commerce.getProductTypes().getProductTypeById(productType|first) %}

        {% if productTypeData and productTypeData.isStructure %}

            {% set queryOrder = 'structure' %}

        {% endif %}
```

**Reply by masiorama** [reply]
Sorry I don't understand your code; what is that `queryOrder` var used for?
What do you put inside `.orderBy()` to target the structure sorting?
edit: let's assume all my product types are structures.

**Reply by masiorama** [reply]
@lukeholder any suggestion? Can you point me to the right direction? Thanks!

**Reply by RitterKnightCreative** [reply]
@masiorama you'd probably have better luck asking on craft stack exchange or on discord as this discussion is about product features, not usage.

But it looks ike you're trying to query for variants, not products. 

From your code, where is ``relatedProducts`` coming from?  If it's from a field, you can already set the order of those products by dragging them around in the field.

If it's from a query, the default sort order for products (and entries) is their order in the CP if you've set it as structure unless you're overriding it via ``orderBy.``

Variants you can also reorder manually in the CP.  You get the product order first then arrange by variant. 

You shouldn't need to do anything special here, just a quick example:

```
{% for p in craft.products.type('whatever').all %}
{% for v in p.variants.all %}
{{ v.description }}<br>
{% endfor %}
{% endfor %}
```

In your case you proobably want to add some parameters to the variant query to make it work the way you want but the product order gets "set" at the product level, not the variant level.

**Reply by masiorama** [reply]
@RitterKnightCreative  you are right, I opened a few minutes ago a question on stackexchange https://craftcms.stackexchange.com/questions/42116/how-to-query-structured-products-commerce-5-2-to-sort-products.

**Reply by masiorama** [reply]
@RitterKnightCreative no luck till now neither on stackexchange, do you have any hint? Thanks

**Reply by RitterKnightCreative** [reply]
@masiorama replied on stack, hope it helps.

**Reply by masiorama** [reply]
> @masiorama replied on stack, hope it helps.

@RitterKnightCreative it's been a while since my last reply but I simply stopped searching for a solution in the last months due to more compelling issues. I found a different solution using `fixedOrder`. More details on SE. Thanks a lot for you help and guidance.
